impeller/renderer/backend/vulkan/gpu_tracer_vk.h

```diff
@@ -0,0 +1,40 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <memory>
+#include "impeller/renderer/backend/vulkan/context_vk.h"
+#include "impeller/renderer/backend/vulkan/vk.h"
+
+namespace impeller {
+
+/// @brief A class that uses timestamp queries to record the approximate GPU
+/// execution time.
+class GPUTracerVK {
+ public:
+  explicit GPUTracerVK(const std::shared_ptr<ContextVK>& context);
+
+  ~GPUTracerVK() = default;
+
+  /// @brief Record the approximate start time of the GPU workload for the
+  ///        current frame.
+  void RecordStartFrameTime();
+
+  /// @brief Record the approximate end time of the GPU workload for the current
+  ///        frame.
+  void RecordEndFrameTime();
+
+ private:
+  void ResetQueryPool(size_t pool);
+
+  const std::shared_ptr<ContextVK> context_;
```

Comment:
I'm not a smart C++ guy. Is the ContextVK owning a shared_ptr to the GPUTracerVK and the GPUTracerVK also having a shared_ptr to the ContextVK going to cause problems? Feels a bit like a cycle. Could use weakptrs...

Reply:
Fixed